Warn users when they run out of space

grote opened the pull request:
This PR takes two approaches for warning the user about out of space situations:

1. try to retrieve actual free space from backup location and warn the user before attempting a backup if this is very low. some hacks are used to get the free space from external storage SAF providers
2. catch exceptions that include hints about out of storage situations and then warn the user as soon as such an exception happens. Note that SAF does not always expose those exceptions to the caller.

Related to #363
Fixes #555

grote committed:
Get free space for storage authority if not available by other means

grote committed:
Implement a storage plugin method to get free space

grote committed:
Store total backup size and show it when restoring

This is done by storing also the APK sizes in addition to the app data size we already store.

grote committed:
Show app restore sets sorted by latest first

grote committed:
Catch out of space errors and show error notification

grote committed:
Don't do backup, if space is already low, warn instead

grote committed:
Don't retry backup on removable storage

t-m-w commented:
> 1. try to retrieve actual free space from backup location and warn the user before attempting a backup if this is very low. some hacks are used to get the free space from external storage SAF providers

Tested this portion with a flash drive, and it looks like it's working fine. Free space is also shown when selecting a backup location, and backup sizes are shown when restoring.

t-m-w commented:
> 2. catch exceptions that include hints about out of storage situations and then warn the user as soon as such an exception happens. Note that SAF does not always expose those exceptions to the caller.

Tested this by setting MIN_FREE_SPACE to -1 and recompiling. Seems to not notice an IOException for ENOSPC (caused by ErrnoException for ENOSPC) and waits a few minutes until timing out with a generic "Backup failed." Full log provided privately.

grote committed:
Catch out of space exception during FullBackup

grote committed:
Catch out 507 HTTP error when using WebDAV

Nextcloud has a bug that lets us write chunked transfers over quota:
https://github.com/nextcloud/server/issues/7993

However, when we upload small files, we can get the proper 507 response and thus detect out of space situations and warn the user about them.

grote commented:
> Seems to not notice an IOException for ENOSPC

This helped me to identify additional out of space situations,we can now warn the user about. Pushed those as extra commits.
